plugins/inputs/influxdb_listener/influxdb_listener.go

```diff
@@ -0,0 +1,395 @@
+}
+
+func (h *InfluxDBListener) Description() string {
+	return "InfluxDB listener"
+}
+
+func (h *InfluxDBListener) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (h *InfluxDBListener) routes() {
+	authHandler := internal.AuthHandler(h.BasicUsername, h.BasicPassword, "influxdb",
+		func(_ http.ResponseWriter) {
+			h.authFailures.Incr(1)
+		},
+	)
+
+	h.mux.Handle("/write", authHandler(h.handleWrite()))
+	h.mux.Handle("/query", authHandler(h.handleQuery()))
+	h.mux.Handle("/ping", h.handlePing())
+	h.mux.Handle("/", authHandler(h.handleDefault()))
+}
+
+func (h *InfluxDBListener) Init() error {
+	tags := map[string]string{
+		"address": h.ServiceAddress,
+	}
+	h.bytesRecv = selfstat.Register("influxdb_listener", "bytes_received", tags)
+	h.requestsServed = selfstat.Register("influxdb_listener", "requests_served", tags)
+	h.writesServed = selfstat.Register("influxdb_listener", "writes_served", tags)
+	h.queriesServed = selfstat.Register("influxdb_listener", "queries_served", tags)
+	h.pingsServed = selfstat.Register("influxdb_listener", "pings_served", tags)
+	h.requestsRecv = selfstat.Register("influxdb_listener", "requests_received", tags)
+	h.notFoundsServed = selfstat.Register("influxdb_listener", "not_founds_served", tags)
+	h.buffersCreated = selfstat.Register("influxdb_listener", "buffers_created", tags)
+	h.authFailures = selfstat.Register("influxdb_listener", "auth_failures", tags)
+	h.routes()
+
+	if h.MaxBodySize.Size == 0 {
+		h.MaxBodySize.Size = defaultMaxBodySize
+	}
+
+	if h.ReadTimeout.Duration < time.Second {
+		h.ReadTimeout.Duration = time.Second * 10
+	}
+	if h.WriteTimeout.Duration < time.Second {
+		h.WriteTimeout.Duration = time.Second * 10
+	}
+
+	return nil
+}
+
+// Start starts the InfluxDB listener service.
+func (h *InfluxDBListener) Start(acc telegraf.Accumulator) error {
+	h.acc = acc
+
+	tlsConf, err := h.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	h.server = http.Server{
+		Addr:         h.ServiceAddress,
+		Handler:      h,
+		ReadTimeout:  h.ReadTimeout.Duration,
+		WriteTimeout: h.WriteTimeout.Duration,
+		TLSConfig:    tlsConf,
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", h.ServiceAddress, tlsConf)
+		if err != nil {
+			return err
+		}
+	} else {
+		listener, err = net.Listen("tcp", h.ServiceAddress)
+		if err != nil {
+			return err
+		}
+	}
+	h.listener = listener
+	h.port = listener.Addr().(*net.TCPAddr).Port
+
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		h.server.Serve(h.listener)
+	}()
+
+	h.log.Infof("Started HTTP listener service on %s", h.ServiceAddress)
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (h *InfluxDBListener) Stop() {
+	deadline := time.Now().Add(10 * time.Second)
+	ctx, cancel := context.WithDeadline(context.Background(), deadline)
+	defer cancel() // go vet requires us to call cancel even though we're waiting for the deadline
+	h.server.Shutdown(ctx)
+	h.wg.Wait()
```

Comment:
So I've read over the docs and I think I've convinced myself that the `wg.Wait` will not hang, however I also thing it's not necessary, since `server.Shutdown(ctx)` is a blocking call, and this can be simplified by removing the waitgroup and it will have the same functionality.

Reply:
I checked this and agree.  I removed the wg